### PR TITLE
[dtensor][debug] added logging module tracing table to file feature

### PR DESCRIPTION
**Pull request by sinhaanshul**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129017
* #128887
* #128729
* __->__ #128721
* #128720

**Summary**
Currently, only way for users to view the module tracing table is to print in the console which could be hard to read. I have added the functionality to comm_debug_mode for a user to log the module tracing table to output.txt file giving the user more options to view module tracing. I have implemented the use case in the module tracing examples. The expected output is shown below for MLPModule tracing:
Screenshot 2024-06-14 at 10 39 07 AM

**Test Plan**
1. torchrun --standalone --nnodes=1 --nproc-per-node=4 torch/distributed/_tensor/examples/comm_mode_features_example.py -e MLP_module_tracing

2. torchrun --standalone --nnodes=1 --nproc-per-node=4 torch/distributed/_tensor/examples/comm_mode_features_example.py -e transformer_module_tracing

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/_tensor/examples/comm_mode_features_example.py [hunk 1]
```diff
@@ -130,17 +130,17 @@ def test_MLPStacked_distributed_sharding_display(self) -> None:
...
-        c10d_functional.all_reduce: 1
+        *c10d_functional.all_reduce: 1
         MLPModule
-            c10d_functional.all_reduce: 1
+            *c10d_functional.all_reduce: 1
             MLPModule.net1
             MLPModule.relu
             MLPModule.net2
-            c10d_functional.all_reduce: 1
+            *c10d_functional.all_reduce: 1
```
Comment (tianyu-l) [inline torch/distributed/_tensor/examples/comm_mode_features_example.py:143]: this indentation seems different from the one in the PR summary?

**Review by tianyu-l — APPROVED** [review]
lgtm!

File: torch/distributed/_tensor/examples/comm_mode_features_example.py (hunk 1 above)
Comment (XilunWu) [reply torch/distributed/_tensor/examples/comm_mode_features_example.py:143]: Yeah I think the one in summary is the newer behavior. Need update in-code comment.

**Review by XilunWu — APPROVED** [review]
LGTM. If the in-code comments need update, you can create a new PR to fix.

**Pull request closed without merging**